Have pathInode.Truncate and friends use the file handle, if available

rfjakob opened the pull request:
If the file handle is specified we should use it instead of
picking one at random.

Fixes #95.

rfjakob commented:
Tests added. Turns out only for fallocate and ftruncate the kernel actually passes in a file handle. For fstat, fchmod and fchown, the added code for checking if the file handle is there is just for future-proofing.

hanwen commented:
this adds API surface, which I'd like to avoid, and using a global variable at that.

Can't you just do 
 
  mount loopback
  f = open("mnt/a")
  mv("orig/a", "orig/b")
  touch("orig/a")
  
 ftruncate(f)

check that b is changed, not a.

rfjakob commented:
That test would fall short i believe, because the kernel always identifies
files by node id. That's not a proof that the file handle was used.

But maybe deleting the file works, will check.

hanwen commented:
but in PathFS, the node ID is transformed back into a path, so you can tell if the FD is used by looking at which file was touched.

rfjakob commented:
Yes, but it tries first to use an open fh

rfjakob commented:
I have not found a way to test this without adding API surface. So to move this forward I have dropped the test and updated the patch. After all, implementing the test had quite a benefit for my understanding of what happens, which I wrote down in the code comments.

I have also added a patch that enables running the integration tests in `fuse/test`, and also fixed the bug that they uncovered immediately.

rfjakob commented:
You have already fixed the mtime bug in ac2190345467cc7bb600f33be326c68af12b8bc0, dropped the patch and rebased to latest master.
